sphinx/ext/intersphinx.py

```diff
@@ -533,17 +534,90 @@ def run(self) -> tuple[list[Node], list[system_message]]:
         assert self.name == self.orig_name.lower()
         inventory, name_suffix = self.get_inventory_and_name_suffix(self.orig_name)
         if inventory and not inventory_exists(self.env, inventory):
-            logger.warning(__('inventory for external cross-reference not found: %s'),
-                           inventory, location=(self.env.docname, self.lineno))
+            self._emit_warning(
+                __('inventory for external cross-reference not found: %r'), inventory
+            )
             return [], []
 
-        role_name = self.get_role_name(name_suffix)
+        domain_name, role_name = self._get_domain_role(name_suffix)
+
         if role_name is None:
-            logger.warning(__('role for external cross-reference not found: %s'), name_suffix,
-                           location=(self.env.docname, self.lineno))
+            self._emit_warning(
+                __('invalid external cross-reference suffix: %r'), name_suffix
+            )
             return [], []
 
-        result, messages = self.invoke_role(role_name)
+        # attempt to find a matching role function
+        role_func: RoleFunction | None
```

Comment:
After sleeping : maybe we can delegate the logic of getting the role function to a specific method which would also include the warnings.

Then, check if it is None, if so return `[], []`, else proceed normally.

Reply:
err, I generally prefer not to introduce to much "indirection" (functions, calling function, calling functions, ...) as it can get annoying to track down the full logic of a function.

But honestly I'm not that bothered in this instance.

Reply:
Actually, it was possibly for extensions developers (like perhaps they want to change this logic).

Reply:
I'm not making it public, I think there should be a strong use case to make any new public methods, and I don't think there is one here at present